File: python/src/runapi/core/pricing.py
"""Live Pricing API resources."""

from __future__ import annotations

from typing import Any, Optional

from .auth import resolve_optional_api_key
from .models import BaseModel, optional, required
from .http_client import HttpClient
from .options import ClientOptions, RequestOptions
from .resource import Resource


class PriceSchedule(BaseModel):
    service = required(str)
    action = required(str)
    model = optional(str)
    pricing_status = required(str)
    catalog_status = required(str)
    currency = required(str)
    billing_unit = optional(str)
    billing_strategy = optional(str)
    unit_price_cents = optional(int)
    input_price_per_1m_cents = optional(int)
    output_price_per_1m_cents = optional(int)
    cache_read_price_per_1m_cents = optional(int)
    cache_write_5m_price_per_1m_cents = optional(int)
    cache_write_1h_price_per_1m_cents = optional(int)
    billing_config = optional()


class PriceScheduleListResponse(BaseModel):
    as_of = required(str)
    price_schedules = required([PriceSchedule])


class PriceQuote(BaseModel):
    service = required(str)
    action = required(str)
    model = optional(str)
    pricing_status = required(str)
    currency = required(str)
    reservation_amount_cents = required(int)
    estimate_basis = required(str)
    as_of = required(str)


class PriceQuoteResponse(BaseModel):
    price_quote = required(PriceQuote)


class Pricing(Resource):
    """Read current price schedules and create reservation estimates."""

    SCHEDULES_ENDPOINT = "/api/v1/price_schedules"
    QUOTES_ENDPOINT = "/api/v1/price_quotes"

    def list_schedules(
        self,
        service: Optional[str] = None,
        action: Optional[str] = None,
        model: Optional[str] = None,
        options: Optional[RequestOptions] = None,
    ) -> Optional[PriceScheduleListResponse]:
        filters = self._compact_params({"service": service, "action": action, "model": model})
        path = self.SCHEDULES_ENDPOINT
        if filters:
            from urllib.parse import urlencode

            path = f"{path}?{urlencode(filters)}"
        return self._request("get", path, options=options, response_class=PriceScheduleListResponse)

    def create_quote(
        self,
        service: str,
        action: str,
        model: Optional[str] = None,
        params: Optional[dict[str, Any]] = None,
        options: Optional[RequestOptions] = None,
    ) -> PriceQuoteResponse:
        body = {"service": service, "action": action, "model": model, "params": params}
        return self._request(
            "post", self.QUOTES_ENDPOINT, body=self._compact_params(body), options=options, response_class=PriceQuoteResponse
        )


class PricingClient(Pricing):
    """Standalone Pricing client with optional API authentication."""

    def __init__(self, api_key: Optional[str] = None, **options: Any) -> None:
        client_options = ClientOptions(api_key=resolve_optional_api_key(api_key), **options)
        super().__init__(client_options.http_client or HttpClient(client_options))

File: ruby/lib/runapi/core/pricing.rb
# frozen_string_literal: true

module RunApi
  module Core
    class Pricing
      include RunApi::Core::ResourceHelpers

      SCHEDULES_ENDPOINT = "/api/v1/price_schedules"
      QUOTES_ENDPOINT = "/api/v1/price_quotes"

      class Schedule < RunApi::Core::BaseModel
        required :service, String
        required :action, String
        optional :model, String
        required :pricing_status, String
        required :catalog_status, String
        required :currency, String
        required :billing_unit, String
        required :billing_strategy, String
        optional :unit_price_cents, Numeric
        optional :input_price_per_1m_cents, Numeric
        optional :output_price_per_1m_cents, Numeric
        optional :cache_read_price_per_1m_cents, Numeric
        optional :cache_write_5m_price_per_1m_cents, Numeric
        optional :cache_write_1h_price_per_1m_cents, Numeric
        required :billing_config, Hash
      end

      class ScheduleListResponse < RunApi::Core::BaseModel
        required :as_of, String
        required :price_schedules, [Schedule]
      end

      class ScheduleNotModifiedResponse < RunApi::Core::BaseModel
        required :not_modified, TrueClass
      end

      class QuoteResponse < RunApi::Core::BaseModel
        required :service, String
        required :action, String
        optional :model, String
        required :pricing_status, String
        required :currency, String
        required :reservation_amount_cents, Numeric
        required :estimate_basis, String
        required :as_of, String
      end

      class QuoteEnvelope < RunApi::Core::BaseModel
        required :price_quote, QuoteResponse
      end

      def initialize(http)
        @http = http
      end

      def list(service: nil, action: nil, model: nil, options: nil)
        query = URI.encode_www_form(compact_params(service:, action:, model:))
        path = query.empty? ? SCHEDULES_ENDPOINT : "#{SCHEDULES_ENDPOINT}?#{query}"
        request_options = (options || RequestOptions.new).dup
        request_options.allow_not_modified = true
        response = @http.request(:get, path, options: request_options)
        payload = response.is_a?(Core::Response) ? response.body : response
        response_class = payload["not_modified"] ? ScheduleNotModifiedResponse : ScheduleListResponse
        result = RunApi::Core::BaseModel.coerce(payload, as: response_class)
        result.with_response_headers(response.response_headers) if response.is_a?(Core::Response)
        result
      end

      def quote(service:, action:, model: nil, params: {}, options: nil)
        response = request(
          :post,
          QUOTES_ENDPOINT,
          body: compact_params(service:, action:, model:, params:),
          options:,
          response_class: QuoteEnvelope
        )
        response.price_quote
      end
    end

    # Standalone live Pricing client with optional API authentication.
    class PricingClient < Pricing
      def initialize(api_key: nil, **options)
        client_options = Core::ClientOptions.new(
          api_key: Core::Auth.resolve_optional_api_key(api_key),
          **options
        )
        @owns_http_client = client_options.http_client.nil?
        super(client_options.http_client || Core::HttpClient.new(client_options))
      end

      # Closes the SDK-created HTTP client. Injected clients remain caller-owned.
      def close
        @http.close if @owns_http_client
      end
    end
  end
end
